### DISCUSSION TITLE
Calling the Telemetry collector from a SPA

### DISCUSSION THREAD

**Discussion by shaunpearsondev** [microsoft/aspire#1914] [Q&A]
I am trying to get telemetry from a react app to show up in the aspire dashboard however I am running into CORS issues

```Access to XMLHttpRequest at 'http://localhost:16151/v1/traces' from origin 'https://localhost:5173' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.```

Is there any way to configure the allowed headers so we can send browser telemetry?

**Comment by davidfowl** [comment]
It’s an interesting scenario for sure but we only support grpc endpoints which don’t work from the browser. I’d also very that most it’ll exporters support this or we might be setting you for failures in deployed environments

**Comment by samsp-msft** [comment]
It would not be rocket science to include a "proxy" api in the host application that would convert from the http form of OTLP to gRPC and send data to the configured telemetry endpoint.

What format is the react app using for its telemetry data?

**Comment by DamianEdwards** [comment]
> It would not be rocket science to include a "proxy" api in the host application that would convert from the http form of OTLP to gRPC and send data to the configured telemetry endpoint.

If by "host application" you mean the AppHost, wouldn't that lead folks down a path that might not be supported by the target environment (@davidfowl's point). Or did you mean in the ASP.NET Core app "hosting" the SPA?

**Comment by samsp-msft** [comment]
Whatever is hosting the SPA - so that the SPA can callback to the host and not have to worry about CORS etc. It also solves the problem of gRPC over http etc.

It could be something like `app.MapReactTelemetry()` the endoint would either:
- make direct OTLP calls based on the environment vars (simpler)
or
- create its own ILogger/Metrics/ActivitySources for the "client" and translate the data it receives to those datasources. The benefit of the latter is that it would be re-emitted via the .NET OTel infra so would work with whatever telemetry backend the app dev had configured, eg Application Insights. The problem would be to distinguish the client results from the server results. To be cleanest, you would probably want 2 instances of OTel registered with DI and I don't think its designed for multiple instances like that.

**Comment by shaunpearsondev** [comment]
I have started to look into something like this using https://www.envoyproxy.io/ and https://www.npmjs.com/package/@opentelemetry/instrumentation in the SPA

I managed to get get everything hosted in the APPHost and exposing endpoints that were available as env vars

REACT => ENVOY => OTEL
Envoy - Docker
React app - EXE (https://github.com/dotnet/aspire-samples/tree/main/samples/AspireWithNode)

Envoy was receiving the data but couldnt quite get it to forward to the OTEL endpoint

Will try and get the experiment pushed to a repo if anyone is interested

**Comment by samsp-msft** [comment]
@JamesNK - what would it take to implement the OTLP/http endpoint in the dashboard? https://github.com/open-telemetry/oteps/blob/main/text/0099-otlp-http.md - its basically the same proto data, but over http/1.1 or h2? The spec says it can even be the same port, disambiguated by http header - can we do that with the standard gRPC support in ASP.NET today?

**Comment by davidfowl** [comment]
It's not something we should prioritize right now, but @shaunpearsondev please experiment and if you have time to research into what other are doing please let us know.

**Comment by wertzui** [comment]
Any news on a HTTP endpoint which will allow instrumentation of SPAs?
I'm currently migration everything to Aspire.
Everything in .Net worked just fine, but I want to instrument my Angular Application too.
I'm now at the point that I get everything injected into the Angular App, can setup everything, but I'm now blocked with the same error.

We really need a possibility to see the traces from our SPAs in Aspire to get the full "experience".

**Reply by wertzui** [reply]
https://github.com/dotnet/aspire/pull/4197 seems to be the solution. Once it is merged, everything should work.

**Reply by davidfowl** [reply]
Except it won't because of CORS. I think we should get some advice from @martinjt about what the typical usage of this is. Making this work outside of dev scenarios means that the otel http endpoint is exposed publicly and without auth?

**Reply by martinjt** [reply]
You shouldn't expose your telemetry backend (in this case Aspire) directly to the internet. This is (partly) the reason that the OpenTelemetry collector exists, to protect your backend platform.

In a production scenario, you want a proxy that limits and doesn't expose your backend keys.

It's fine in dev, but production, you shouldn't be exposing your Aspire dashboard to the internet. You can use the collector as that gateway, but even then I'd consider hosting things separately as you'll want things like Rate limiting, etc.

The way I recommend deploying this stuff is to hook into the reverse proxy you're using for splitting traffic on a single domain, and map `/v1/*` to the otel collector and have that forward the traffic to your telemetry backend.

Keep in mind that there's a limited amount of span/log capacity in aspire, and frontend otel will generate a lot of data, perhaps more than your backend depending on the size.

**Reply by wertzui** [reply]
Isn't Aspire all about the Dev experience?

As @martinjt said, in Prod we would have another setup using proxies, collectors and a backend which actually stores the OTEL data.

But in a Dev scenario all of these don't exist and are replaced by Aspire, so it should be capable of ingesting the data and displaying it on the Dashboard.

**Reply by martinjt** [reply]
The point is that you should deploy the collector locally too, to replicate what's happening when you got to prod.

There's an aspire package for the collector.

**Reply by JamesNK** [reply]
@wertzui What library are you using in the Angular app to export telemetry? I looked at the opentelemetry-js and it seems to have packages for exporting to gRPC - https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages

@davidfowl @martinjt I think we want to favor telemetry just working in development environment and have documentation/guidance for best practices in production.

That means:

* Support HTTP OTLP protocol
* Enable CORS by default when the dashboard is run in development. Thoughts on security implications?

**Reply by davidfowl** [reply]
I think the first step here is to make a working example that works both locally in aspire and also when deployed to ACA. Lets get an end to end working before we make any decisions as how we support this in the product natively.

**Reply by wertzui** [reply]
Im currently using these libraries:
```json
"@opentelemetry/api": "^1.8.0",
"@opentelemetry/auto-instrumentations-web": "^0.39.0",
"@opentelemetry/context-zone": "^1.24.1",
"@opentelemetry/exporter-trace-otlp-http": "^0.51.1",
"@opentelemetry/sdk-trace-web": "^1.24.1",
```

So the exporter is not the gRPC exporter, but the HTTP exporter, because browsers do not (yet) support gRPC.

**Reply by JamesNK** [reply]
Ok. I was wondering whether grpc-web is an option, but probably not.

`exporter-trace-otlp-http` is OTLP + HTTP + JSON.

We'll probably initially only support OTLP + HTTP + PROTO. I looked around and that's supported with `exporter-trace-otlp-proto`. The npm package is different but the client configuration is the same. The end result is OTLP data sent as protobuf instead of JSON: https://www.npmjs.com/package/@opentelemetry/exporter-trace-otlp-proto

**Reply by martinjt** [reply]
My worry is that if the dashboard supports this, people will override that in production.

I think solid documentation, and easy mode collector that just makes it work would be a great solution here, and also introduces people to the right way to do it.

The library I created that will add a default collector hooked up to the dashboard etc. Would be all they need to make it work.

**Reply by davidfowl** [reply]
The dashboard will support HTTP as well as gRPC but I don't think we should add support for CORS until we understand what the end to end looks like (deployed to ACA as an example). It seems like telemetry should be sent to the backend serving the client side app (a BFF) and that application should be responsible for proxying the data to the otlp endpoint in a secure manner. 

We try our best to make sure the modeling of the application is as close to what it would be when deployed so we can preserve as much as the experience as possible:

Local:

![image](https://github.com/dotnet/aspire/assets/95136/f0aa34b1-9542-491e-bbba-1b822beba4bf)

Deployed:

![image](https://github.com/dotnet/aspire/assets/95136/722e6c49-25c1-4e69-84a6-882b4273eb96)

**Reply by martinjt** [reply]
I wouldn't recommend that as an approach for production apps, I'd split the /v1/* endpoints at the ingress rather than the app.

Thw application shouldn't be participating in the trace and adding an additional load on the app in that scenario, which adding an endpoint would do, you'd be creating a trace for every batch of spans that the frontend sends.

**Reply by davidfowl** [reply]
Can you show me in a diagram what you'd recommend?

**Reply by martinjt** [reply]
![example aspire deploy](https://github.com/dotnet/aspire/assets/1699587/cdbc51f5-e5fb-40fe-bf38-9a56f8ccd907)

Something like that, I'm not sure what you can do with the ingress for ACA right now, but for a production app, you'd probably have Frontdoor infront of that anyway to manage throttling/rate limiting/abuse anyway. In that scenario, I'd replace "ingress" with "frontdoor".

**Reply by davidfowl** [reply]
Got it! So the the assumption is that something is configuring HTTP ingress rules to route to the collector with all of the security setup so that the ingress can talk forward this traffic to the collector (with the right auth).  Can you elaborate on what the /v1/* is? I'd like to get to a concrete sample that we can deploy and experiment with.

Is the ingress forwarding the OTLP HTTP requests directly with a minimal transform to add the appropriate auth headers etc etc?

**Reply by martinjt** [reply]
The otel signals default to:

* `/v1/traces`
* `/v1/logs`
* `/v1/metrics`

The frontend could be producing any or all of these signals. Later there will be profiles, and other things, but they all sit under `/v1` for HTTP (gRPC is obviously different).

The ingress would do nothing to the request over than routing. In this scenario, the collector can't have auth, since any kind of API key would be pointless as the client would need it, or just hitting the ingress endpoint would add it anyway. For frontend observability, it has to be open and anonymous access, otherwise you can only trace when someone has logged in.

The collector "could" be configured to have a separate receiver for frontend telemetry and backend telemetry, but that's probably beyond the scope of this right now.

**Reply by martinjt** [reply]
sidenote, if you want frontend, you can't use the ACA managed "collector", since it's not able to be made accessible outside the environment.

**Reply by JamesNK** [reply]
A publicly exposed collector creates issues around people abusing it. I think you would want to lock it down on the server so only telemetry for certain applications is accepted. For example, an allow list that allows telemetry for the browser app and rejects someone maliciously creating telemetry for a backend app. And should incoming requests require a session key in the cookie or header that the site generates?

There are probably solutions to make this work well, Google Analytics has been around for decades. But I wonder if ingress + collector be hardened enough to make it a secure option?

**Reply by wertzui** [reply]
How this this handled in Azure Monitor / Application Insights?
When using the official libraries, it will directly connect to a publicly exposed endpoint using a connection string that anyone can read in their browser.

**Reply by martinjt** [reply]
In those scenarios, where a vendor exposes their backend to the Internet for clientside analytics data, they're built specific for it.

They have:
* throttling
* rate limiting
* burst protection
* behaviour analysis

They're built differently in a lot of ways to the ones where you can trust that the auth is private.

For the dashboard/backend to be public, there will be things like that to consider. It's things that the Clientside SIG in OpenTelemetry are thinking about. The answer will be collector configurations that are hardened.

**Comment by JamesNK** [comment]
Updates on this scenario:

* OTLP HTTP will be supported in 8.1.
* However, CORS is not yet supported. Browser requests to the dashboard will be blocked. Issue to add support: https://github.com/dotnet/aspire/issues/4917
